rust/agents/relayer/src/msg/pending_message.rs

```diff
@@ -0,0 +1,377 @@
+    pub destination_mailbox: CachingMailbox,
+    /// Origin chain database to verify gas payments.
+    pub origin_db: HyperlaneDB,
+    /// Used to construct the ISM metadata needed to verify a message from the
+    /// origin.
+    pub metadata_builder: BaseMetadataBuilder,
+    /// Used to determine if messages from the origin have made sufficient gas
+    /// payments.
+    pub origin_gas_payment_enforcer: Arc<GasPaymentEnforcer>,
+    /// Hard limit on transaction gas when submitting a transaction to the
+    /// destination.
+    pub transaction_gas_limit: Option<U256>,
+    pub metrics: MessageSubmissionMetrics,
+}
+
+/// A message that the submitter can and should try to submit.
+#[derive(new)]
+pub struct PendingMessage {
+    pub message: HyperlaneMessage,
+    ctx: Arc<MessageContext>,
+    #[new(default)]
+    submitted: bool,
+    #[new(default)]
+    submission_data: Option<Box<SubmissionData>>,
+    #[new(default)]
+    num_retries: u32,
+    #[new(value = "Instant::now()")]
+    last_attempted_at: Instant,
+    #[new(default)]
+    next_attempt_after: Option<Instant>,
+}
+
+/// State for the next submission attempt generated by a prepare call.
+struct SubmissionData {
+    metadata: Vec<u8>,
+    gas_limit: U256,
+}
+
+impl Debug for PendingMessage {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        // intentionally leaves out ctx
+        let now = Instant::now();
+        let last_attempt = now.duration_since(self.last_attempted_at).as_secs();
+        let next_attempt = self
+            .next_attempt_after
+            .map(|a| {
+                if a >= now {
+                    a.duration_since(now).as_secs()
+                } else {
+                    0
+                }
+            })
+            .unwrap_or(0);
+        write!(f, "PendingMessage {{ num_retries: {}, since_last_attempt_s: {last_attempt}, next_attempt_after_s: {next_attempt}, message: {:?} }}",
+               self.num_retries, self.message)
+    }
+}
+
+impl PartialEq for PendingMessage {
+    fn eq(&self, other: &Self) -> bool {
+        self.num_retries == other.num_retries
+            && self.message.nonce == other.message.nonce
+            && self.message.origin == other.message.origin
+    }
+}
+
+impl Eq for PendingMessage {}
+
+#[async_trait]
+impl PendingOperation for PendingMessage {
+    fn domain(&self) -> &HyperlaneDomain {
+        self.ctx.destination_mailbox.domain()
+    }
+
+    #[instrument]
+    async fn prepare(&mut self) -> PendingOperationResult {
+        make_op_try!(|| self.on_retry());
+
+        if !self.is_ready() {
+            trace!("Message is not ready to be submitted yet");
+            return PendingOperationResult::NotReady;
+        }
+
+        // If the message has already been processed, e.g. due to another relayer having
+        // already processed, then mark it as already-processed, and move on to
+        // the next tick.
+        let is_already_delivered = op_try!(
```

Comment:
Both ways made sense to me but went for making a slight change to the serial submitter to check if it has been submitted after calling prepare since that was already kinda supported.

Reply:
Actually there is another reason why calling with lag is not great.

If we call with lag and it says that it is not there but it is at latest, we end up in an edge case where we attempt to submit a transaction that will fail because it is a duplicate.

By not using lag and instead moving it to the confirm queue, we can handle it just as we do any other transaction from then on out.

Reply:
> If we call with lag and it says that it is not there but it is at latest, we end up in an edge case where we attempt to submit a transaction that will fail because it is a duplicate.

Right, this is what I was referring to when I mentioned the risk of submitting unecessarily

> By not using lag and instead moving it to the confirm queue, we can handle it just as we do any other transaction from then on out.

But we aren't moving it to the confirm queue, we are dropping it

Reply:
Not anymore, current code returns `PendingOperationResult::Success` after setting submitted to true, and then that ends up pushing it through to the confirm queue.